telegram/ext/dispatcher.py

```diff
@@ -448,57 +512,58 @@ def update_persistence(self, update=None):
             update (:class:`telegram.Update`, optional): The update to process. If passed, only the
             corresponding ``user_data`` and ``chat_data`` will be updated.
         """
-        if self.persistence:
-            chat_ids = self.chat_data.keys()
-            user_ids = self.user_data.keys()
-
-            if isinstance(update, Update):
-                if update.effective_chat:
-                    chat_ids = [update.effective_chat.id]
-                else:
-                    chat_ids = []
-                if update.effective_user:
-                    user_ids = [update.effective_user.id]
-                else:
-                    user_ids = []
-
-            if self.persistence.store_bot_data:
-                try:
-                    self.persistence.update_bot_data(self.bot_data)
-                except Exception as e:
-                    try:
-                        self.dispatch_error(update, e)
-                    except Exception:
-                        message = 'Saving bot data raised an error and an ' \
-                                  'uncaught error was raised while handling ' \
-                                  'the error with an error_handler'
-                        self.logger.exception(message)
-            if self.persistence.store_chat_data:
-                for chat_id in chat_ids:
+        with self._update_persistence_lock:
+            if self.persistence:
+                chat_ids = list(self.chat_data.keys())
+                user_ids = list(self.user_data.keys())
```

Comment:
What is this `list()` change for?

Reply:
KeyViews or whatever it's called change, when the dict changes, which is inconvenient here, as we're going to loop over the keys. If we receive an update from an unknown user/chat while doing so, `chat_ids` will change and we get an error like [this](https://t.me/pythontelegrambotgroup/363178) one.
The use of `list()` decouples `chat_ids` from `self.chat_data` and just considers the keys present at call-time. The new entry is ignored and will be handled in the next call to `update_persistence`, which the update from the unknown user triggers.

Reply:
Excellent, please add it as a comment on the code.